himan-scripts/traffic_weather_warning.lua

```diff
@@ -0,0 +1,354 @@
+          level = level(HPLevelType.kHeight, level_height),
+          param = param(param_name),
+          forecast_type = ftype,
+          read_previous_forecast_if_not_found = false
+  }
+
+  data = luatool:FetchWithArgs(o)
+  return data
+end
+
+function fetch_radiation_data(step, level_height, prod_time, ftype, is_sum)
+  local o = {
+    forecast_time = prod_time,
+    level = level(HPLevelType.kHeight, level_height),
+    forecast_type = ftype
+  }
+
+  local agg_duration, multiplier
+  if step < 83 then
+    agg_duration = time_duration("01:00")
+    multiplier = 1
+  elseif step < 137 then
+    agg_duration = time_duration("03:00")
+    multiplier = 3
+  else
+    agg_duration = time_duration("06:00")
+    multiplier = 3
+  end
+
+  o.param = param('RADGLO-WM2', aggregation(HPAggregationType.kAverage, agg_duration), processing_type())
+
+  -- Fetch radiation data
+  if is_sum and step < 83 then
+    -- Fetch for the current and two prior time steps for summing
+    local radglo_sum = fetch_multiple_radiation(o, prod_time)
+    return radglo_sum
+  else
+    -- Fetch single time step data and apply multiplier if needed
+    local radglo_data = luatool:FetchWithArgs(o)
+    if is_sum then
+      for i = 1, #radglo_data do
+        radglo_data[i] = multiplier * radglo_data[i]
+      end
+    end
+    return radglo_data
+  end
+end
+
+-- Helper function to fetch data for multiple time steps and sum them
+function fetch_multiple_radiation(o, prod_time)
+  local my_time = forecast_time(prod_time)  
+
+  local radglo_sum = {}
+  for j = 0, 2 do
+    if j > 0 then
+      my_time:GetValidDateTime():Adjust(HPTimeResolution.kHourResolution, -1)
+    end
+    o.forecast_time = my_time
+    local radglo_data = luatool:FetchWithArgs(o)
+    for i = 1, #radglo_data do
+      radglo_sum[i] = (radglo_sum[i] or 0) + radglo_data[i]
+    end
+  end
+  return radglo_sum
+end
+
+
+ftype_d = forecast_type(HPForecastType.kDeterministic)
+ftype_c = forecast_type(HPForecastType.kEpsControl, 0)
+
+vire_time = get_prod_time('VIRE_PREOP', 287)
+ecmwf_time = get_prod_time('ECGMTA', 240)
+meps_time = get_prod_time('MEPSMTA', 260)
+
+-- if past 66 hours, priority 2 data is ec, not meps
+meps_step = tonumber(meps_time:GetStep():Hours())
+pri2_time = nil
+pri2_ftype = nil
+
+if meps_step < 66 then 
+  pri2_time = meps_time
+  pri2_ftype = ftype_c
+else 
+  pri2_time = ecmwf_time
+  pri2_ftype = ftype_d
+end
+
+tk = fetch_parameter('T-K', 2, vire_time, ftype_d)
+ff = fetch_parameter('FF-MS', 10, vire_time, ftype_d)
+nlm = fetch_parameter('NLM-PRCNT', 0, vire_time, ftype_d)
+prec = fetch_parameter('PRECFORM2-N', 0, vire_time, ftype_d)
+tdk = fetch_parameter('TD-K', 2, vire_time, ftype_d)
+
+radglo1 = fetch_radiation_data(meps_step, 0, pri2_time, pri2_ftype, false)
+radglo3 = fetch_radiation_data(meps_step, 0, pri2_time, pri2_ftype, true)
+
+ec_step = tonumber(ecmwf_time:GetStep():Hours())
+
+-- if ec step over 85, the steps increases and validation time can't be shifted
```

Comment:
Kaksi  kysymystä:
1) Miksi käytetään aina EC:n lumisadetta eikä MEPS:n lumisadetta silloin kun ennustepituus on < 66h?
2) Mitä tarkoittaa tuo että "validation time can't be shifted", ja miksi siitä johtuen pitkillä ennustepituuksilla snr6 ja snr12 ovat nollia?


Yksi huomio:

Makrossa on katsottu lumisadetta eteen- ja taaksepäin; tässä koodissa otetaan vain kertymä (eli katsotaan vain taaksepäin). Mietin että seuraako tästä se että lua-koodi ei anna varoitusta silloin jos lumisade on tulossa, eli käytännössä Himanin antama varoitus olisi viivästynyt ajallisesti makron varoitukseen verrattuna.


Kun makrossa on näin: `SumT(-1, +1, par264_EC)`, niin sen voisi kääntää Himaniin niin että hakisi  `SN-3-MM` tunnilta +1. Samaten `SumT(-3, +3, par264_EC)` voisi olla `SN-6-MM` tunnilta +3 (se ei vastaisi täysin samaa koska se olisi 6h kertymä eikä 7h mutta olisi varmasti riittävän lähellä) ja `SumT(-6, +6, par264_EC` olisi `SN-12-MM` tunnilta +6.

Reply:
Tässä kohdassa ajatuksena on siirtää lumisadekertymää eteenpäin, jotta varoitus ei viivästy. Riveillä 309-312 kertymiä siirretään move_valid_time()-funktiolla 1, 2 tai 5 tuntia eteenpäin. Näin tehdään vain jos EC-step on alle 85, koska sen jälkeen stepit ovat yli tunnin, joten (ainakin luulisin) valid-aikaa ei voi siirtää eteenpäin, koska näitä aikoja ei enää ole. Vaikka stepit tarkistetaan EC-ajoista, data tulee silti ennen MEPS 66 steppiä MEPS:stä pri2_time-muuttujan mukaan.

1. Ennen 66h käytetään MEPS lumisadetta. Haettu lumiparametri riippuu pri2_time-muuttujasta, joka asetetaan rivillä 288 MEPS-stepin mukaan.
2. Tämä johtuu oletuksestani, että steppien kasvaessa valid-aikoja puuttuu eikä niitä pysty enää siirtämään tulevaisuuteen, kuten ennen EC-steppiä 85.

Reply:
Olen hiukan hidas niin kertaan vielä. Eli pri2_ftype määrittää miltä ennustetyypiltä lumisade haetaan. pri2_time määrittää ajan mutta sehän voi olla (teoriassa ainakin) sama molemmilla tuottajilla. Esim 00 -analyysi löytyy sekä ec:ltä että mepsiltä, mutta näistä tuottajista vain toinen on deterministinen ja toinen parven kontrolli ja siksi haku menee oikeaan paikkaan.

Mutta jäin vielä miettimään tuota validtimen siirtämistä. Tämä varoitus-data lasketaan viren päälle joka on taas rakennettu ec:n ja mepsin päälle, eli kaikki ajat joille varoitus lasketaan on myös olemassa muilta tuottajilta. Eli jos vaikka ennustetunti olisi 101, niin edellinen tunti on 98 ja seuraava 104. Tässä tapauksessa miksi et ottaisi dataa seuraavasti:
* snr3 tunnilta 101
* snr6 tunnilta 104
* snr12 tunnilta 109

Reply:
Joo siis tosiaan se pri2_ftype, myös vaikuttaa haettuun ennusteeseen! Eli juuri, kuten sanoit pri2_time ja pri2_ftype määrittelee datan tuottajan.

Joo noinhan se kannattaa tehdä! Jotenkin ajattelin, että yhden tai kahden tunnin siirto eteenpäin ei  onnistu, ni en edes miettiny niitä seuraavia ennusteita :D Siirrän niitä ton verran eteenpäin. Kun steppi on 6h ja ennustetunti esimerkiksi 161, niin kertymäajat voisi olla:

- snr3 tunnilta 161
- snr6 tunnilta 167 (menee vähän pitkälle, mutta 161 kertymä saattais olla vähän myöhässä veroituksen kannalta?)
- snr12 tunnilta 167

Reply:
Tuo kuulostaa järkevältä kompromissilta. snr3 ja snr12 ottavat kuitenkin huomioon menneen sateen kun snr6 tässä tapauksessa keskittyy vain tulevaan sateeseen.